public/js/components/util/ManagedTree.js

```diff
@@ -0,0 +1,58 @@
+"use strict";
+
+const React = require("react");
+const Tree = React.createFactory(require("../../lib/tree"));
+
+let ManagedTree = React.createClass({
+  propTypes: Tree.propTypes,
+
+  getInitialState() {
+    return { expanded: new WeakMap(),
+             focusedItem: null };
+  },
+
+  setExpanded(item, expanded) {
+    const e = this.state.expanded;
+    e.set(item, expanded);
+    this.setState({ expanded: e });
+
+    if (expanded && this.props.onExpand) {
+      this.props.onExpand(item);
+    } else if (!expanded && this.props.onCollapse) {
+      this.props.onCollapse(item);
+    }
+  },
+
+  focusItem(item) {
+    if (this.state.focused !== item) {
+      this.setState({ focusedItem: item });
+
+      if (this.props.onFocus) {
+        this.props.onFocus(item);
+      }
+    }
+  },
+
+  render() {
+    const { expanded, focusedItem } = this.state;
+
+    const props = Object.assign({}, this.props, {
+      isExpanded: item => expanded.get(item),
```

Comment:
I think some code comments here would be really helpful on why this ManagedTree is needed and how some of the handlers are flowing through. It's a little hard to un-pack what is going on here, even though I know you mentioned to me briefly that you needed this. I'm finding the props on here to be particularly confusing. It somewhat breaks the metaphor in my head of how props get passed in from above, because this component is essentially shadowing the the parent's props. The only two properties that are overwritten are the `onFocus` and `renderItem`. Could these be different names, more explicit names to make this clearer?

I don't know if things could be refactored a little bit to make it clearer what's doing what... It could just be a matter of being more vigorous with comments for some context.